presto-main/src/main/java/com/facebook/presto/operator/aggregation/noisyaggregation/NoisySumAggregationUtils.java

```diff
@@ -0,0 +1,147 @@
+
+public class NoisySumAggregationUtils
+{
+    private NoisySumAggregationUtils()
+    {
+    }
+
+    public static void updateNoisySumState(NoisySumState state, double value, double noiseScale, Double lower, Double upper, Long randomSeed)
+    {
+        state.setNoiseScale(noiseScale);
+        checkNoiseScale(state.getNoiseScale());
+
+        state.setCount(state.getCount() + 1);
+
+        if (randomSeed == null) {
+            state.setNullRandomSeed(true);
+        }
+        else {
+            state.setNullRandomSeed(false);
+            state.setRandomSeed(randomSeed);
+        }
+
+        checkLowerUpper(lower, upper);
+        double clippedValue = value;
+        if (lower != null && upper != null) {
+            state.setNullLowerUpper(false);
+            state.setLower(lower);
+            state.setUpper(upper);
+
+            clippedValue = clip(value, lower, upper);
+        }
+
+        state.setSum(state.getSum() + clippedValue);
+    }
+
+    public static void combineNoisySumStates(NoisySumState state, NoisySumState otherState)
+    {
+        state.setNoiseScale(state.getNoiseScale() > 0 ? state.getNoiseScale() : otherState.getNoiseScale()); // noise scale should be > 0
+        checkNoiseScale(state.getNoiseScale());
+        if (state.isNullLowerUpper()) {
+            state.setNullLowerUpper(otherState.isNullLowerUpper());
+            state.setLower(otherState.getLower());
+            state.setUpper(otherState.getUpper());
+        }
+
+        state.setCount(state.getCount() + otherState.getCount());
+
+        if (state.isNullRandomSeed()) {
+            state.setNullRandomSeed(otherState.isNullRandomSeed());
+            state.setRandomSeed(otherState.getRandomSeed());
+        }
+
+        state.setSum(state.getSum() + otherState.getSum());
+    }
+
+    public static void writeNoisySumStateOutput(NoisySumState state, BlockBuilder out)
+    {
+        if (state.getCount() == 0) {
+            out.appendNull();
+            return;
+        }
+
+        Random random = new SecureRandom();
+        if (!state.isNullRandomSeed()) {
+            random = new Random(state.getRandomSeed());
+        }
+        // add noise
+        double noiseSdv = state.getNoiseScale();
+        double noise = random.nextGaussian() * noiseSdv;
+        double trueSum = state.getSum();
+        double noisySum = trueSum + noise;
+        // fix sign
+        double noisySumFixedSign = fixSignNoisySum(noisySum, state.getLower(), state.getUpper());
+
+        DOUBLE.writeDouble(out, noisySumFixedSign);
+    }
+
+    /**
+     * Clip value to [lower, upper] range
+     */
+    public static double clip(double value, double lower, double upper)
+    {
+        return Math.max(lower, Math.min(upper, value));
+    }
+
+    /**
+     * Given a range [lower, upper], making sure that the sign of noisy sum is consistent with the sign of the bound.
+     */
+    public static double fixSignNoisySum(double noisySum, double lower, double upper)
```

Comment:
I'm on the fence about whether these sorts of things should be always-on behavior. This kind of post-processing is probably often desirable but occasionally undesirable. In particular, this results in biased (not zero mean) noise addition for values close to zero. Depending on the situation, this could be surprising. For example, if you were to select `NOISY_SUM_GAUSSIAN(x, noiseScale, 0, 100) GROUP BY y`, then roll up (i.e., sum) the resulting values, the noisy rollup would overstate the true sum on average.

Reply:
I am on the fence too

Reply:
Removed fixSign part